driver/csiplugin/nodeserver.go

```diff
@@ -48,60 +50,92 @@ const nodePublishMethodSymlink = "SYMLINK"
 const isOpenShiftCluster = "IS_OpenShift"
 const mountPathLength = 6
 
+// A map for locking/unlocking a target path for NodePublish/NodeUnpublish
+// calls. The key is target path and value is a boolean true in case there
+// is any NodePublishVolume or NodeUnpublishVolume request in progress for
+// the target path.
+var nodePublishUnpublishLock map[string]bool
+
+// a mutex var used to make sure certain code blocks are executed by
+// only one goroutine at a time.
+var mutex sync.Mutex
+
+func lock(targetPath string, ctx context.Context) bool {
+	mutex.Lock()
+	defer mutex.Unlock()
+
```

Comment:
What would happen if we get multiple request for Node Publish with different target path at same time. One go routine will be holding the above map variable and in mean time other go routines need to wait till it is unlocked. I hope probably there might be a dead lock situation. Can we try some test cases around this and check.

Reply:
CSI driver behavior wise which I have seen in UT, when multiple requests for NodePublish come for different paths to a driver instance on a node, only one NodePublish request is handled by the driver and then only next one comes, and this is the behavior I have seen with or without the locking mechanism introduced here.
Code wise, there won't be any deadlock as the mutex based locking and unlocking is done only while putting or deleting the entry in the map to avoid put/delete by multiple requests for the same path. The NodePublish and Unpublish locking and unlocking is based on that entry only.

Reply:
This PR works fine when 100 pods were created/deleted in a loop.